### ISSUE TITLE
Should bower_components be in/below the web root?

### ISSUE THREAD

**Issue by mpdude** [bower/bower#1487]
Reading the docs, I am missing some guidance on where to best put the `bower_components` folder. In particular, should it be below or outside the webservers document/web root? 

If a package contains for example CSS files and additional assets (say, `background-image`s with relative `url()`s), placing it below the web root makes usage really simple as you can use the CSS as-is. If you already have some CSS-concatenation-and-minification tasks in place, they will probably include URL rebasing steps that will work for this setup already.

Not having `bower_components` below the web root definetly requires you to come up with some build process. URL rewriting might become more difficult and you would also have to maintain a list of all images, fonts and the like that you would like to copy over to web. 

On the other hand, Bower packages might contain Sass or Less files that need not be publicly available, or might even be private/internal and must not be exposed. So keeping `bower_components` with the source and deploying only final CSS as part of the build process is easier than fiddling around with webserver access control rules...

How do you folks handle this? Are there some recommendations we could put into the docs somewhere?

**Comment by topping** [comment]
@benschwarz: as a new Bower user, I've been struggling with this as well. Your comment helps a lot. Why not automate it though? Shouldn't ``bower.json`` have pointers to the location(s) of the usable assets? Developers could continue putting stuff wherever their personality suits them. These pointers should somehow be usable by the build automation.

Coming from a Maven background, a lot of this was possible because the integration points all lived under one roof. Here, adapters have to be developed and maintained for {name all build systems like Grunt}. 

Grateful for any thoughts you might have, I feel like I am missing something big but fear that I'm not.

**Comment by mpdude** [comment]
I am unable to find the reference right now, but a while ago I checked and found that `main` does not even have a clear definition and does not matter to Bower itself (?). I'm afraid that unless there's a clear recommendation what should go in there it will be hard to build re-usable tools on top of it.

**Comment by topping** [comment]
Yes, I felt the definitions in http://bower.io/docs/creating-packages/ were vague and open to interpretation. Maybe we have our answer?

The best way to define these is to create sensible defaults and allow redefinition per package if necessary. Then get a few projects to agree and use them. If the standards take hold because tools can leverage the new properties to reduce developer effort, the community will adopt them.

**Comment by waynebloss** [comment]
My personal workflow for bower up to now is to simply use it as a build downloader and version tracker. I only put `bower.json` in my repo and I ignore `bower_components`, even for website projects. After downloading a bower package, I then manually copy it's files wherever I want and they are stored in the repo in their new location.

Today I was investigating the possibility of changing my workflow to let bower manage more things. (*e.g. Start using [wiredep](taptapship/wiredep) and other tools*), but honestly it seems like a lot of work for very little gain. If I used CDNs more, there might be one benefit there but I'm still not sure it's worth the trouble because all that automation code now has to be maintained. I don't think there is a good middle ground. Either fully automate or don't.

I think I'm going to keep my simple workflow for now. My `gulpfile.js` is a lot smaller this way (even more-so when I'm using my [staticbuild](https://github.com/devoptix/staticbuild) program to run the dev server and gulp doesn't have to watch things). The end result is that gulp only has to build and deploy my sites.

**Comment by derrekbertrand** [comment]
So after buying a template, integrating their gulpfile into my Laravel project, installing Bower, fixing dependencies, and going through package by package fixing the relative URLs; I realized that I had a bunch of demo files, useless images, and source LESS/SASS, that I'm serving up for the world to snoop in. Apparently, this is actually how it works, and there isn't a way to automate without adding another tool, and even then they might not support `bower-install` or whatever.

As a back-end developer, I just run `composer whatever-the-heck`, possibly add a line to one of my configs, and even more rarely - run a single-line command to tell it to finish installing itself. Ya'll front-end folks need to get on our level. :laughing: 

I remembered this article when I was messing with bower. Originally it was kinda funny; now it's all too real for me:
https://hackernoon.com/how-it-feels-to-learn-javascript-in-2016-d3a717dd577f
